cpp/src/join/quadtree_point_to_nearest_polyline.cu

```diff
@@ -30,101 +30,118 @@
-  cudf::mutable_column_device_view out_poly_index,
-  cudf::mutable_column_device_view out_distance)
+inline __device__ thrust::tuple<uint32_t, uint32_t> get_quad_poly_and_local_point_indices(
+  uint32_t const global_index, uint32_t const *point_offsets, uint32_t const *point_offsets_end)
 {
-  // each block processes a quadrant
-  auto const quad_pos = blockIdx.x + gridDim.x * blockIdx.y;
-  auto const quad_idx = quad_idxs[quad_pos];
-
-  auto const poly_begin = poly_idx_offsets[quad_pos];
-  auto const poly_end =
-    quad_pos < num_poly_idx_offsets - 1 ? poly_idx_offsets[quad_pos + 1] : num_poly_idxs;
-
-  auto const num_points = quad_lengths.element<uint32_t>(quad_idx);
-
-  for (auto tid = threadIdx.x; tid < num_points; tid += blockDim.x) {
-    // each thread loads its point
-    auto const point_pos = quad_offsets.element<uint32_t>(quad_idx) + tid;
-    auto const point_id  = point_indices.element<uint32_t>(point_pos);
-    auto const px        = point_x.element<T>(point_id);
-    auto const py        = point_y.element<T>(point_id);
-    auto nearest_poly_id = static_cast<uint32_t>(-1);
-    T distance_squared   = std::numeric_limits<T>::max();
-
-    for (uint32_t poly_id = poly_begin; poly_id < poly_end; poly_id++)  // for each polyline
-    {
-      auto const poly_idx   = poly_idxs[poly_id];
-      auto const ring_begin = poly_offsets.element<uint32_t>(poly_idx);
-      auto const ring_end   = poly_idx < poly_offsets.size() - 1
-                              ? poly_offsets.element<uint32_t>(poly_idx + 1)
-                              : poly_points_x.size();
-      auto const ring_len = ring_end - ring_begin;
-
-      for (auto point_idx = 0; point_idx < ring_len; point_idx++)  // for each line
-      {
-        auto const i0  = ring_begin + ((point_idx + 0) % ring_len);
-        auto const i1  = ring_begin + ((point_idx + 1) % ring_len);
-        auto const x0  = poly_points_x.element<T>(i0);
-        auto const y0  = poly_points_y.element<T>(i0);
-        auto const x1  = poly_points_x.element<T>(i1);
-        auto const y1  = poly_points_y.element<T>(i1);
-        auto const dx0 = px - x0, dy0 = py - y0;
-        auto const dx1 = px - x1, dy1 = py - y1;
-        auto const dx2 = x1 - x0, dy2 = y1 - y0;
-        auto const d0 = dx0 * dx0 + dy0 * dy0;
-        auto const d1 = dx1 * dx1 + dy1 * dy1;
-        auto const d2 = dx2 * dx2 + dy2 * dy2;
-        auto const d3 = dx2 * dx0 + dy2 * dy0;
-        auto const r  = d3 * d3 / d2;
-        auto const d  = d3 <= 0 || r >= d2 ? min(d0, d1) : d0 - r;
-        if (d < distance_squared) {
-          distance_squared = d;
-          nearest_poly_id  = poly_idx;
-        }
-      }
-    }
-
-    out_point_index.element<uint32_t>(point_pos) = point_pos;
-    out_poly_index.element<uint32_t>(point_pos)  = nearest_poly_id;
-    out_distance.element<T>(point_pos)           = sqrt(distance_squared);
-  }
+  // Calculate the position in "point_offsets" that `global_index` falls between.
```

Comment:
This function is very similar to `compute_poly_and_point_indices` functor in quadtree_point_in_polygon. Can you factor out a common utility?

Reply:
Done in https://github.com/rapidsai/cuspatial/pull/362/commits/1bc89b4eab9d7033b6177097b45981939e794865.